### DISCUSSION TITLE
Download Attachments via Link

### DISCUSSION THREAD

**Discussion by Brrrradley** [Budibase/budibase#11350] [Help] [answered]
I have an attachments field that can hold multiple attachmemts.  These attachments subsequently need to be presented for download.  When using a table presentation, the list of attachments gets truncated (+ 4 more ) and are no longer accessible.  As below.

![attachment1](https://github.com/Budibase/budibase/assets/101414789/d1cd14aa-3a5c-442d-ba57-055f91e3d7cd)

I have played around with nested repeater blocks and array fields to iterate over all of the files in an attachment field and present them differently as links (see below).  This is fine, but I can't seem to be able to download the file using this approach, which is my desired outcome.

![attachment2](https://github.com/Budibase/budibase/assets/101414789/adb05351-3675-45b6-bb68-40c5a2c5d0b9)

Any ideas?  Either displaying all attachments in tabular format (eliminating the +4 more) or allowing a file to be downloaded via link component would be great. 

Thanks.

**Comment by ConorWebb96** [comment, marked as answer]
Hey @Brrrradley,

I think i've achieved what you are after with some minor JS plus the use of the repeater and embed component. Below is the component tree as well as the JS code used within the binding to achieve the desired effect.

**Component tree**
<img width="269" alt="Screenshot 2023-07-27 at 07 52 56" src="https://github.com/Budibase/budibase/assets/126772285/4990d04d-6437-4068-8be2-214d2c13656a">

**JS Code**
<img width="1466" alt="Screenshot 2023-07-27 at 07 53 44" src="https://github.com/Budibase/budibase/assets/126772285/5d8460b8-d110-4798-891f-7fc6947d9488">

```
const attachments = $("New Repeater block.test.attachments");

if(attachments) {
  let html = '';
    for (const attachment of attachments) {
      html += `<a href="${attachment.url}" download>${attachment.name}</a><br>`;
    }

  return html;
}
return 'No attachments here';
```

**Output**
<img width="1416" alt="Screenshot 2023-07-27 at 07 54 39" src="https://github.com/Budibase/budibase/assets/126772285/f5e4412e-d671-4ce4-987e-5ff08d2b46c1">


I hope this makes sense and is helpful, if you have any questions around this let me know.

**Reply by Brrrradley** [reply]
Thanks @ConorWebb96  -- this worked exactly as needed.

**Reply by melohagan** [reply]
There is also the [Link component](https://docs.budibase.com/docs/link) which may work with the repeater block as well, but this solution is nice because it makes it convenient to add a message if there are no attachments! 

Great answer! 👍

**Reply by NicolasFabre81** [reply]
Hello @melohagan ,
I just tried this approach with the link component but can't make it work.
The URL I get is  :
https://bb-stg.fr.mybbhost.lcl/app_dev_519ffbdd536b4325b53066675b3b67bb#/files/signed/prod-budi-app-assets/app_519ffbdd536b4325b53066675b3b67bb/attachments/488f5f4a-24bf-497e-8397-c11b44b719ca.xlsx?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=y6s2WUi6rJCktMo3fEfJEHoEPyoHnGew%2F20231002%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20231002T094458Z&X-Amz-Expires=3600&X-Amz-Signature=75d81911ee212199b76ffe35e446b11c0455e901e5246b5a6055a5d70284f216&X-Amz-SignedHeaders=host

It opens a new tab but I get the app default page, my XSLX file does not download.

**Reply by SaschaValentin** [reply]
Hello @NicolasFabre81, I have the same problem, even if I use "key" attribute to reference the PDF file in the bindings. Strange thing is it shows the right link (https://dataapps.mediacomvalentin.de/app/app_b46f6a00e4934fe29b11d903a253c2fb/attachments/8d2adad3-60c7-4c00-bc21-24ef86fc699a.pdf) to the file but it doesn´t open it. Instead the homepage of the app is opened.

I hope someone can explain what we are doing wrong.

**Comment by spitoutzi** [comment]
Hi! How to download files with the original file names ( like foto1.jpg) and not like 8d44rf-frefw7e87rw6-fewu437f.jpg?

**Reply by melohagan** [reply]
You can't at the moment. There's a related issue here: https://github.com/Budibase/budibase/issues/11301

**Reply by spitoutzi** [reply]
I managed to implement it as follows.
Via Handlebars (using the download="{{name}}" parameter):
```
{{# forEach Repeater.Table.file }}
<br>
<a href="{{url}}" target="_blank" class="download" download="{{name}}">Download</a>
</br>
{{/forEach}}
```

**Reply by melohagan** [reply]
Awesome, thanks for sharing!

**Comment by papka1974** [comment]
I used Navigate to action providing downloadable file URL and everything worked. Maybe it's a bit easier solution.
